### DISCUSSION TITLE
How to monitor the change of focus of a LineEdit?

### DISCUSSION THREAD

**Discussion by Horbin-Magician** [slint-ui/slint#3908] [Q&A]
I have seen that "Added focus-changed-event callback to FocusScope." in the [Changelog](https://github.com/slint-ui/slint/blob/master/CHANGELOG.md#changelog) of Slint 1.3.0. But I want to do something when the focus of a LineEdit changed, Is there any way we can do that now? Perhaps I should wait for https://github.com/slint-ui/slint/issues/3811 to be implemented?

**Comment by ogoffart** [comment]
What are you trying to achieve? Is it to pop up a virtual keyboard? What platform are you using?

**Reply by Horbin-Magician** [reply]
I'm trying to achieve a search tool in Win11. Pressing Ctrl+F displays a search box coantaining a LineEdit, which hides when the LineEdit loses focus.

**Reply by ogoffart** [reply]
I see.  Then #3811 might not help you since that is only for writing your own platform.
I think what you want is maybe a callback in the Window itself when the window gets de-activated, would that help?

**Reply by Horbin-Magician** [reply]
Yes, I remember that you give me a good trick to achieve this purpose. It like:
```Rust
pure callback lose_focus_trick(bool) -> bool;
always-on-top: lose_focus_trick(input.has-focus);
```
This is do not work well all the time, and sometimes it is called when no change has occurred.

**Comment by Tsuiban** [comment]
Adding on another use case, I want to do a my_line_edit.select_all() whenever my LineEdit gains focus.  I can do this with a TextEdit via the FocusScope.  However, LineEdit doesn't accept a FocusScope due to not allowing it to have any children.

I tried surrounding it with a Rectangle and putting the FocusScope there, but that causes lots of other problems such as being unable to tab out of the field to the next field.

Closest I've gotten was this:

```
GridBox {
        spacing: 0;
        Row {
            Text { text: "Start Time: "; }
            Rectangle {
                forward-focus: start_time_focus_handler;
                start_time_focus_handler := FocusScope {
                    focus-changed-event() => {
                        if self.has_focus {
                            st.select-all();
                        }
                    } }

                st := LineEdit {
                    placeholder-text: "hh:mm:ss";
                    accepted => { root.data-changed(); }
                    width: 100px;
                }
            }
        }
```

**Comment by FischLu** [comment]
Any news of this issue??

**Reply by ogoffart** [reply]
With Slint 1.8 you can do:

```slint
 LineEdit {
    changed has-focus => { ... }
 }
```

**Reply by FischLu** [reply]
Hi, thanks for the info, however, it doesn't work, the app just crash with the following error
```
thread 'main' panicked at /Users/xxx/.cargo/registry/src/index.crates.io-6f17d22bba15001f/i-slint-compiler-1.8.0/llr/expression.rs:676:89:
  called `Option::unwrap()` on a `None` value
```

I just add 
```
LineEdit {
      width: 50px;
      input-type: number;
      changed has-focus => { debug("hello") }
  }
```
to a slint offical template, If I remove the line `changed has-focus => { debug("hello") }` , the app works without any issues. Actually, if I add just `changed has-focus => {}`, the app crashes as well. By the way, in LineEdit documentation there is no `changed` callback mentioned, does this callback support all widgets??

**Reply by FischLu** [reply]
tried more about `changed` callback, `changed text => { debug("hello") }` works, but `has-focus` doesn't, is it a bug?

**Reply by hunger** [reply]
Yes, that looks like a bug. Could you please file an issue, ideally with a short snippet of code reproducing the problem? That would help a lot!

**Reply by FischLu** [reply]
Okay

**Comment by felipetesc** [comment]
Even though that's not what you asked, what about using TextEdit and configuring its height, wrap, x, and y, (other stuff too), so you mimic LineEdit ?

**Reply by FischLu** [reply]
but TextEdit doesn't have onfocus callback as well
